### DISCUSSION TITLE
ImageMagick 6.9.10-68 Q16 x86_64 - Exception 420: no decode delegate for this image format

### DISCUSSION THREAD

**Discussion by kul-bhushan-prasad** [ImageMagick/ImageMagick#2297] [Help]
After the upgrade of library, my code is failing, please help and let me know what is the issue and equal ant syntax so that i can use to fix the issue with verion 6.7.8*:

Code:
$x = $qImage->Resize(geometry=>"144x188", filter=>'Cubic');
    $log->error( "Border resize fail=$x") if $x;

my $quadImage=Image::Magick->new(magick=>'jpg');
    $x = $quadImage->Read(filename=>$one_white_pixel);
    $log->error("Couldn't read white pixel=$x") if $x;
    $quadImage->Resize(geometry=>'200x200');
    $log->error("Couldn't set background and size=$x") if $x;
 $x = $quadImage->Composite(image=>$qImage, gravity=>"Center");
    $log->error("Composite the photo onto the background=$x") if $x;

 # Use the Transform() method to scale an image and automatically apply an intelligent 
    # sharpening filter that will give better results than Scale( ) in most cases.
    $x = $tnImage->Resize(geometry=>$sizes{tn}{w} . 'x' . $sizes{tn}{h}, filter=>'Cubic');
    $log->error("TN Resize Result=$x") if "$x";

Exception:
07-16-20 13:57:01 PID=7381 error:Border resize fail=Exception 410: invalid filter value `Cubic' @ error/Q16HDRI.xs/XS_Image__Magick__Q16HDRI_Mogrify/7735
07-16-20 13:57:01 PID=7381 error:Couldn't read white pixel=Exception 420: no decode delegate for this image format `/tmp/magick-l-haKu6i' @ error/constitute.c/ReadImage/544
07-16-20 13:57:01 PID=7381 error:Couldn't set background and size=Exception 420: no decode delegate for this image format `/tmp/magick-l-haKu6i' @ error/constitute.c/ReadImage/544
07-16-20 13:57:01 PID=7381 error:Composite the photo onto the background=Exception 410: no images defined `Composite' @ error/Q16HDRI.xs/XS_Image__Magick__Q16HDRI_Mogrify/7620
07-16-20 13:57:01 PID=7381 error:TN Resize Result=Exception 410: invalid filter value `Cubic' @ error/Q16HDRI.xs/XS_Image__Magick__Q16HDRI_Mogrify/7735

**Comment by snibgo** [comment]
IM v6.7.8-9 is very very very old. I suggest you upgrade to a more recent version.

**Reply by kul-bhushan-prasad** [reply]
Done, still failing

**Comment by fmw42** [comment]
IM 6.7.8.9 is known to have bugs. I suggest you upgrade to a current version. IM is now at 6.9.11.24

**Comment by kul-bhushan-prasad** [comment]
With 6.9.10-68 version getting these exceptions.

Code:
my $quadImage=Image::Magick->new(magick=>'jpg');
$x = $quadImage->Read(filename=>$one_white_pixel);
$log->error("Couldn't read white pixel=$x") if $x;
$quadImage->Resize(geometry=>'200x200');
$log->error("Couldn't set background and size=$x") if $x;
$x = $quadImage->Composite(image=>$qImage, gravity=>"Center");
$log->error("Composite the photo onto the background=$x") if $x;

exception:
07-20-20 14:22:30 PID=34458 debug:ImageMagick 6.9.10-68 Q16 x86_64 2019-11-13 https://imagemagick.org
07-20-20 14:22:30 PID=34458 debug:kuprasad: 1085 x 1446 r= 0.75
07-20-20 14:22:31 PID=34458 error:Couldn't read white pixel=Exception 420: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564
07-20-20 14:22:31 PID=34458 error:Couldn't set background and size=Exception 420: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564
07-20-20 14:22:31 PID=34458 error:Composite the photo onto the background=Exception 410: no images defined `Composite' @ error/Q16HDRI.xs/XS_Image__Magick__Q16HDRI_Mogrify/7620

**Reply by fmw42** [reply]
What is the value of `$one_white_pixel`? Does it contain @? If so, you likely need to edit your ImageMagick policy.xml file to give permission to use @ to read files.  Why are you trying to resize a 1 pixel image. Why not just create your 200x200 size image of the same color?

**Reply by kul-bhushan-prasad** [reply]
Hi, I tried to print, the value of $one_white_pixel is complete image path like /opt/httpd/cdcapp-nprd1/root/htdocs/data-shared/dir/photos/one_white_pixel.jpg and it is valid.

it does not contains @. The latest log is

07-20-20 16:30:09 PID=42229 debug:ImageMagick 6.9.10-68 Q16 x86_64 2019-11-13 https://imagemagick.org

07-20-20 16:30:09 PID=42229 debug:kuprasad: 1085 x 1446 r= 0.75
07-20-20 16:30:09 PID=42229 debug:one_white_pixel /opt/httpd/cdcapp-nprd3-101/root-stage3_xxphoto-5115/htdocs/data-shared/dir/photos/one_white_pixel.jpg
07-20-20 16:30:09 PID=42229 error:Couldn't read white pixel=Exception 420: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564
07-20-20 16:30:09 PID=42229 error:Couldn't set background and size=Exception 420: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564
07-20-20 16:30:09 PID=42229 error:Composite the photo onto the background=Exception 410: no images defined `Composite' @ error/Q16HDRI.xs/XS_Image__Magick__Q16HDRI_Mogrify/7620

**Reply by fmw42** [reply]
Is one_white_pixel.jpg valid?  Can you read it in ImageMagick command line directly, such as with `identify -verbose one_white_pixel.jpg`?

**Reply by kul-bhushan-prasad** [reply]
Sure here you go.

identify -verbose /opt/httpd/cdcapp-nprd3-101/root-stage3_xxphoto-5115/htdocs/data-shared/dir/photos/one_white_pixel.jpg

Image: /opt/httpd/cdcapp-nprd3-101/root-stage3_xxphoto-5115/htdocs/data-shared/dir/photos/one_white_pixel.jpg
Format: JPEG (Joint Photographic Experts Group JFIF format)
Mime type: image/jpeg
Class: DirectClass
Geometry: 1x1+0+0
Resolution: 72x72
Print size: 0.0138889x0.0138889
Units: PixelsPerInch
Colorspace: sRGB
Type: Grayscale
Base type: Undefined
Endianess: Undefined
Depth: 8-bit
Channel depth:
red: 8-bit
green: 8-bit
blue: 8-bit
Channel statistics:
Pixels: 1
Red:
min: 254 (0.996078)
max: 254 (0.996078)
mean: 254 (0.996078)
standard deviation: 0 (0)
kurtosis: -3
skewness: 0
entropy: 0
Green:
min: 254 (0.996078)
max: 254 (0.996078)
mean: 254 (0.996078)
standard deviation: 0 (0)
kurtosis: -3
skewness: 0
entropy: 0
Blue:
min: 254 (0.996078)
max: 254 (0.996078)
mean: 254 (0.996078)
standard deviation: 0 (0)
kurtosis: -3
skewness: 0
entropy: 0
Image statistics:
Overall:
min: 254 (0.996078)
max: 254 (0.996078)
mean: 254 (0.996078)
standard deviation: 0 (0)
kurtosis: -3
skewness: 0
entropy: 0
Colors: 1
Histogram:
1: (254,254,254) #FEFEFE srgb(254,254,254)
Rendering intent: Perceptual
Gamma: 0.454545
Chromaticity:
red primary: (0.64,0.33)
green primary: (0.3,0.6)
blue primary: (0.15,0.06)
white point: (0.3127,0.329)
Background color: white
Border color: srgb(223,223,223)
Matte color: grey74
Transparent color: black
Interlace: None
Intensity: Undefined
Compose: Over
Page geometry: 1x1+0+0
Dispose: Undefined
Iterations: 0
Compression: JPEG
Quality: 89
Orientation: TopLeft
Properties:
date:create: 2013-03-27T20:56:19+00:00
date:modify: 2006-04-20T22:55:12+00:00
dc:format: image/jpeg
exif:ColorSpace: 1
exif:DateTime: 2006:02:24 16:28:18
exif:ExifOffset: 164
exif:Orientation: 1
exif:PixelXDimension: 1
exif:PixelYDimension: 1
exif:ResolutionUnit: 2
exif:Software: Adobe Photoshop CS Windows
exif:thumbnail:Compression: 6
exif:thumbnail:JPEGInterchangeFormat: 302
exif:thumbnail:JPEGInterchangeFormatLength: 553
exif:thumbnail:ResolutionUnit: 2
exif:thumbnail:XResolution: 72/1
exif:thumbnail:YResolution: 72/1
exif:XResolution: 72/1
exif:YResolution: 72/1
icc:copyright: Copyright (c) 1998 Hewlett-Packard Company
icc:description: sRGB IEC61966-2.1
icc:manufacturer: IEC http://www.iec.ch
icc:model: IEC 61966-2.1 Default RGB colour space - sRGB
jpeg:colorspace: 2
jpeg:sampling-factor: 2x2,1x1,1x1
signature: e86ee2d99c1767a8000330e9d12515cfc3170a6a22deb3d64e2cce6eaab15e62
tiff:Orientation: 1
tiff:ResolutionUnit: 2
tiff:XResolution: 72/1
tiff:YResolution: 72/1
xap:CreateDate: 2006-02-24T16:28:18-07:00
xap:CreatorTool: Adobe Photoshop CS Windows
xap:MetadataDate: 2006-02-24T16:28:18-07:00
xap:ModifyDate: 2006-02-24T16:28:18-07:00
xapMM:DocumentID: adobe:docid:photoshop:e560421a-a58c-11da-b640-8d8d7fb455b9
xmp:ColorSpace: 1
xmp:PixelXDimension: 1
xmp:PixelYDimension: 1
Profiles:
Profile-8bim: 2028 bytes
Profile-exif: 861 bytes
Profile-icc: 3144 bytes
Profile-xmp: 6103 bytes
Artifacts:
filename: /opt/httpd/cdcapp-nprd3-101/root-stage3_xxphoto-5115/htdocs/data-shared/dir/photos/one_white_pixel.jpg
verbose: true
Tainted: False
Filesize: 12748B
Number pixels: 1
Pixels per second: 403B
User time: 0.000u
Elapsed time: 0:01.002
Version: ImageMagick 6.9.10-68 Q16 x86_64 2019-11-13 https://imagemagick.org

**Reply by fmw42** [reply]
I do not see anything obviously bad about that image. So I do not know why you are getting that error message. Perhaps a bad version of ImageMagick? Sorry, I do not know your API, only the command line?  Does the same processing in command line fail?

**Reply by kul-bhushan-prasad** [reply]
Hi, thanks for help..
I am using ImageMagick 6.9.10-68 and code is mentioned below.

my $quadImage=Image::Magick->new(magick=>'jpg');
$x = $quadImage->Read(filename=>$one_white_pixel);
$log->error("Couldn't read white pixel=$x") if $x;
$quadImage->Resize(geometry=>'200x200');
$log->error("Couldn't set background and size=$x") if $x;
$x = $quadImage->Composite(image=>$qImage, gravity=>"Center");
$log->error("Composite the photo onto the background=$x") if $x;

The problem is, it's failing in server itself.

**Reply by fmw42** [reply]
I am sorry, I do not know that API software code.

**Reply by kul-bhushan-prasad** [reply]
It's ok, May be this can help.

1) The code works with /usr/bin/perl and (v5.16.3)
2) The code fails with /usr/local/bin/perl (v5.8.9)

We do not know what is missing in Perl v5.8.9 and what additionally we have to install to get rid of this issue.

**Reply by fmw42** [reply]
Why don't you simply upgrade your version of Perl? Sound to me like Perl is the issue? You have a very old version of Perl that fails, that may be flawed or corrupt. So just upgrade Perl?

Perhaps a bad or corrupt version of PerlMagick? Have you tried re-installing PerlMagick?

See https://imagemagick.org/script/perl-magick.php. Perhaps that may help?

**Reply by kul-bhushan-prasad** [reply]
Thanks, It's shared sever with some customized configurations We will check on it and update. Thanks again.
